Bl 2790 custom license

hatton opened the pull request:
The original problem was that Bloom couldn't differentiate between a custom license and a empty license. As we did recently with topics, I took the opportunity to switch from a bi-direction binding to a uni-directional system, which made the code cleaner and easier to unit test and reason about. 

Then I found that Bloom had license handling scattered throughout the code base, between languages, etc., much of it duplicating each other. So I gathered all that into a single c# class and gave it unit tests.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/bloombooks/bloomdesktop/885)
<!-- Reviewable:end -->

JohnThomson commented:
Reviewed 22 of 22 files at r1.
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[src/BloomBrowserUI/bookEdit/js/bloomEditing.js, line 237 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885#-K4Tden8qp2SPRynqzJ3:-K4Tden8qp2SPRynqzJ4:-755453031) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/f3280af881b3e7fd5b0627b2d2db6cf2e3383481/src/BloomBrowserUI/bookEdit/js/bloomEditing.js#L237)):*
Is there really still something to do here?

---

*[src/BloomExe/Book/Book.cs, line 793 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885#-K4TeJmXo1Sav_am8aei:-K4TeJmXo1Sav_am8aej:765390257) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/f3280af881b3e7fd5b0627b2d2db6cf2e3383481/src/BloomExe/Book/Book.cs#L793)):*
I haven't studied this in depth, but it feels good to me that we shouldn't be updating the files while generating a preview DOM, which seems to be what the previous location of this code was trying to ensure. Is there a good reason to change it?

---

*[src/BloomExe/Book/BookCopyrightAndLicense.cs, line 15 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885#-K4TfAPCOAmjDJtw4h57:-K4TfAPCOAmjDJtw4h58:-763409816) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/f3280af881b3e7fd5b0627b2d2db6cf2e3383481/src/BloomExe/Book/BookCopyrightAndLicense.cs#L15)):*
'has' is redundant

---

*[src/BloomExe/Book/BookCopyrightAndLicense.cs, line 137 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885#-K4ThAv40l91R56jULtT:-K4ThAv40l91R56jULtU:1805602087) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/f3280af881b3e7fd5b0627b2d2db6cf2e3383481/src/BloomExe/Book/BookCopyrightAndLicense.cs#L137)):*
It feels like these two lines belong in the 'no attribute' branch, since the attribute branch is not concerned with lang attributes or element content.

Especially if that's true, I'm wondering whether this code would be more readable with distinct methods for content and attribute storage...the little common bit at the top could become a 3rd method? Not something I feel very strongly about, but worth asking.

---

*[src/BloomExe/Book/BookCopyrightAndLicense.cs, line 166 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885#-K4TiXOCwfSElJXTcRSb:-K4TiXOCwfSElJXTcRSc:1491457580) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/f3280af881b3e7fd5b0627b2d2db6cf2e3383481/src/BloomExe/Book/BookCopyrightAndLicense.cs#L166)):*
This somehow reads as if this method is going to DO something that affects all books without licenses. Perhaps I'm confused because this method doesn't really even seem to be about default licences...it will create the appropriate image file for any CC license, whether it's the default CC-BY or not. Has part of this comment gotten misplaced?

No, I see later that you moved it unchanged. Still a good chance to improve things.

---

*[src/BloomExe/Book/BookInfo.cs, line 368 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885#-K4TjwKi-73Ac_WrXC3R:-K4TjwKi-73Ac_WrXC3S:470633100) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/f3280af881b3e7fd5b0627b2d2db6cf2e3383481/src/BloomExe/Book/BookInfo.cs#L368)):*
Since this class already has a field _metadata, which is a BookMetaData, NOT the CreativeCommons.Metadata object passed here, I think you need a name that will prevent the two from being confused...maybe SetLicenseMetadata?

---

*[src/BloomExe/Book/BookInfo.cs, line 383 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885#-K4Tl7a9NtuwDHEw4M0b:-K4Tl7a9NtuwDHEw4M0c:954610673) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/f3280af881b3e7fd5b0627b2d2db6cf2e3383481/src/BloomExe/Book/BookInfo.cs#L383)):*
Not sure where this code came from, but if we obfuscate in the BookInfo itself, and thus in meta.json, where is the original stored so we can get it back when they download?

---

*[src/BloomExe/Book/HtmlDom.cs, line 786 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885#-K4TlfrHrHMgUv4zGygX:-K4TlfrHrHMgUv4zGygY:265686109) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/f3280af881b3e7fd5b0627b2d2db6cf2e3383481/src/BloomExe/Book/HtmlDom.cs#L786)):*
As we discussed the other day, this business of deleting things in a foreach loop over SafeSelectNodes may call for a ToList or ToArray so we aren't iterating over the collection while modifying the underlying document.

---

*[src/BloomExe/Edit/EditingModel.cs, line 1347 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885#-K4Tm_HoUnFJ_cQGQdGN:-K4Tm_HoUnFJ_cQGQdGO:1197674538) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/f3280af881b3e7fd5b0627b2d2db6cf2e3383481/src/BloomExe/Edit/EditingModel.cs#L1347)):*
ChangeLicenseMetaData? ChangeBookLicenseMetaData? Since there is a class actually called BookMetaData (which this is NOT changing) I find the name confusing. (In fact I wish the CC class had a less generic name, like LicenseMetaData.) (I see the method had this name in its original location, but we're trying to improve things here...)

---

*[src/BloomExe/Edit/EditingModel.cs, line 1350 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885#-K4Tn-WEa-4fx_qq123G:-K4Tn-WEa-4fx_qq123H:-389235251) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/f3280af881b3e7fd5b0627b2d2db6cf2e3383481/src/BloomExe/Edit/EditingModel.cs#L1350)):*
let' => let's, if you can be bothered.

---

*[src/BloomExe/Edit/EditingView.cs, line 235 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885#-K4TnHlgzmrpkrOP1k6n:-K4TnHlgzmrpkrOP1k6o:1858568981) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/f3280af881b3e7fd5b0627b2d2db6cf2e3383481/src/BloomExe/Edit/EditingView.cs#L235)):*
Not even for other people's templates? But I suppose this means it's up to the template designer whether to make CC a default, which may be a nice flexibility to have.

---

*[src/BloomTests/Book/HtmlDomTests.cs, line 352 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885#-K4TrBeDqDeyROH-4kdE:-K4TrBeDqDeyROH-4kdF:220461352) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/f3280af881b3e7fd5b0627b2d2db6cf2e3383481/src/BloomTests/Book/HtmlDomTests.cs#L352)):*
Being very picky, I only see two there to begin with (apart from the one that is something else and NOT removed)

---

*[src/BloomTests/Book/HtmlDomTests.cs, line 410 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885#-K4TrciWyfu-3M3h-FoR:-K4TrciWyfu-3M3h-FoS:1605833697) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/f3280af881b3e7fd5b0627b2d2db6cf2e3383481/src/BloomTests/Book/HtmlDomTests.cs#L410)):*
Is it worth verifying that it actually made a data div to put the new setting into?

---

*[src/BloomTests/Book/HtmlDomTests.cs, line 425 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885#-K4TrzhsYOLKAI8-a-XI:-K4TrzhtZ3wwJSDFkQ29:1040572467) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/f3280af881b3e7fd5b0627b2d2db6cf2e3383481/src/BloomTests/Book/HtmlDomTests.cs#L425)):*
Is it worth testing that it can successfully add a new language where some variations on 'foo' already exist? Also that it can add a brand new setting 'foo' to a data div that already has others?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885)*
<!-- Sent from Reviewable.io -->

hatton commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[src/BloomBrowserUI/bookEdit/js/bloomEditing.js, line 237 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885#-K4Tden8qp2SPRynqzJ3:-K4UwVAClcY-dn8LiHpn:1107378424) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/f3280af881b3e7fd5b0627b2d2db6cf2e3383481/src/BloomBrowserUI/bookEdit/js/bloomEditing.js#L237)):*
nope. thanks.

---

*[src/BloomExe/Book/Book.cs, line 793 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885#-K4TeJmXo1Sav_am8aei:-K4UwpgT1asKG6CnsuiI:2034789047) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/f3280af881b3e7fd5b0627b2d2db6cf2e3383481/src/BloomExe/Book/Book.cs#L793)):*
If this method, "bringBookUpToDate()", is being run on a preview dom, then that's the level at which we'd need to address your concern. Since this is inside of that method, it seems proper to make the change on the book.

---

*[src/BloomExe/Book/BookCopyrightAndLicense.cs, line 137 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885#-K4ThAv40l91R56jULtT:-K4UxwBcw-8DTsaS891L:-2003183651) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/f3280af881b3e7fd5b0627b2d2db6cf2e3383481/src/BloomExe/Book/BookCopyrightAndLicense.cs#L137)):*
Yes, that would be more readable in there. Done.

---

*[src/BloomExe/Book/BookCopyrightAndLicense.cs, line 166 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885#-K4TiXOCwfSElJXTcRSb:-K4Uyk79B-NY0W1i2Gef:-817051327) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/f3280af881b3e7fd5b0627b2d2db6cf2e3383481/src/BloomExe/Book/BookCopyrightAndLicense.cs#L166)):*
Yes, that comment is bizarre. thanks for catching it.

---

*[src/BloomExe/Book/BookInfo.cs, line 368 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885#-K4TjwKi-73Ac_WrXC3R:-K4Uz56V23wH62K-JMo5:-799582260) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/f3280af881b3e7fd5b0627b2d2db6cf2e3383481/src/BloomExe/Book/BookInfo.cs#L368)):*

Yes, good call. I've made the change here, and but the "metadata" name is still overloaded. We'd have to change it in lipalaso to really fix it. From there, it looks like a good name. It is "ClearShare.Metadata" where ClearShare is my never-really-developed gran scheme for a library that handles all IP stuff.

---

*[src/BloomExe/Book/BookInfo.cs, line 383 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885#-K4Tl7a9NtuwDHEw4M0b:-K4V-0QES5dPFWy6-7k2:2132561068) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/f3280af881b3e7fd5b0627b2d2db6cf2e3383481/src/BloomExe/Book/BookInfo.cs#L383)):*
It came from Book.cs. I agree, it's weird.  All I did here was take the opportunity to move it out of the over-burdened Book class and into BookInfo where it seemed to belong better.

---

*[src/BloomExe/Book/HtmlDom.cs, line 786 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885#-K4TlfrHrHMgUv4zGygX:-K4V-eY_kE_hoJNPveyH:-499146009) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/f3280af881b3e7fd5b0627b2d2db6cf2e3383481/src/BloomExe/Book/HtmlDom.cs#L786)):*
Done.

---

*[src/BloomExe/Edit/EditingModel.cs, line 1347 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885#-K4Tm_HoUnFJ_cQGQdGN:-K4V-pjZntWW__-Ws6DF:978600303) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/f3280af881b3e7fd5b0627b2d2db6cf2e3383481/src/BloomExe/Edit/EditingModel.cs#L1347)):*
Good, thanks.

---

*[src/BloomExe/Edit/EditingModel.cs, line 1350 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885#-K4Tn-WEa-4fx_qq123G:-K4V-uIDMx-X11rHoM8F:301519394) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/f3280af881b3e7fd5b0627b2d2db6cf2e3383481/src/BloomExe/Edit/EditingModel.cs#L1350)):*
I long for the day when a reviewer can simply click in the text and make the proposed change, then the author merely approves it, Google Docs style. 

Fixed.

---

*[src/BloomExe/Edit/EditingView.cs, line 235 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885#-K4TnHlgzmrpkrOP1k6n:-K4V0B7HMPzh73Ws_dPq:-1908262855) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/f3280af881b3e7fd5b0627b2d2db6cf2e3383481/src/BloomExe/Edit/EditingView.cs#L235)):*
You can't see it here in the diff, but this section is actually vestigial now, and needs to be removed. 

---

*[src/BloomTests/Book/HtmlDomTests.cs, line 352 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885#-K4TrBeDqDeyROH-4kdE:-K4V7UZVsNnPEtc0h40y:-499146009) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/f3280af881b3e7fd5b0627b2d2db6cf2e3383481/src/BloomTests/Book/HtmlDomTests.cs#L352)):*
Done.

---

*[src/BloomTests/Book/HtmlDomTests.cs, line 410 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885#-K4TrciWyfu-3M3h-FoR:-K4V7r5PWVEukh-2vqMX:629479174) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/f3280af881b3e7fd5b0627b2d2db6cf2e3383481/src/BloomTests/Book/HtmlDomTests.cs#L410)):*
It *is* worth it! Added new test.

---

*[src/BloomTests/Book/HtmlDomTests.cs, line 425 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885#-K4TrzhsYOLKAI8-a-XI:-K4V7zxoqkNzNRoMZlA9:-47978324) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/f3280af881b3e7fd5b0627b2d2db6cf2e3383481/src/BloomTests/Book/HtmlDomTests.cs#L425)):*
Yah know, sometimes I think this is why people don't do as much refactoring as I do... they fear that they will be obligated to make dozens of tests for scenarios that previously were tested only indirectly :-)

Added SetBookSetting_AddANewVariationToAnExistingKey_Added()

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885)*
<!-- Sent from Reviewable.io -->

JohnThomson commented:
LGTM (with one minor quibble)

---

Reviewed 8 of 8 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[src/BloomExe/Edit/EditingModel.cs, line 1347 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885#-K4Tm_HoUnFJ_cQGQdGN:-K4YCT8udOSH4O861ijR:-379982450) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/f3280af881b3e7fd5b0627b2d2db6cf2e3383481/src/BloomExe/Edit/EditingModel.cs#L1347)):*
It's slightly unfortunate that this method now uses just "License" while elsewhere you used "LicenseAndCopyright". I'd be happy with either, but would prefer them to be the same.
Shorter than LicenseAndCopyright but more accurate than just license would be Change/SetClearShareMetaData.
There are also some GetLicenseMetadata methods around...be nice to have everything that relates to ClearShare.MetaData as consistent as possible.

---

*[src/BloomTests/Book/HtmlDomTests.cs, line 425 \[r1\]](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885#-K4TrzhsYOLKAI8-a-XI:-K4YJ8t63NIZgZ3nd-Lp:-1169615860) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/f3280af881b3e7fd5b0627b2d2db6cf2e3383481/src/BloomTests/Book/HtmlDomTests.cs#L425)):*
Yes, it's unfortunate that there's often no way of knowing which of the problems I see are remnants of earlier work where things have already improved.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/885#-:-K4YLR-otSTvtxU1N6La:-1188459031)*
<!-- Sent from Reviewable.io -->
